### PR TITLE
Cleanup input checks for modifications

### PR DESCRIPTION
**Pull request by markuspf**
This an alternative to #11629 and a potential fix for BTS-56 in a structured way; we should be checking whether inputs to transaction methods (insert, update, replace, remove, upsert) are valid; 

In this PR, I intend to check that a passed `_key` is valid. For this I use the collection's `KeyGenerator` that validates that a given key string is valid according to https://www.arangodb.com/docs/3.6/appendix-glossary.html#document-key in the `Local` variant of the operation. Arguably this is a bit late (as we should really be checking this earlier in the input path).

I removed code that seemed to collapse a string `something/somethingelse` input to `somethingelse`, but there didn't seem to be a way to get this input through to this input path at all.

This is still incomplete, I have only changed `update` and `remove` before I ran out of time for this week. 

I'd like to see what Jenkins says about this change.

### CODE REVIEW

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/10056/

File: arangod/Transaction/Methods.cpp [hunk 1]
```diff
@@ -1350,10 +1351,11 @@ Future<OperationResult> transaction::Methods::modifyLocal(std::string const& col
   auto workForOneDocument = [this, &operation, &options, &collection,
                              &resultBuilder, &cid, &previous,
                              &result](VPackSlice const newVal, bool isBabies) -> Result {
-    Result res;
-    if (!newVal.isObject()) {
-      res.reset(TRI_ERROR_ARANGO_DOCUMENT_TYPE_INVALID);
-      return res;
+
+    auto validatedKey = transaction::helpers::validatedOperationInputDocumentKey(*collection, newVal);
```
Comment (mchacki) [inline arangod/Transaction/Methods.cpp:1355]: This only allows objects

File: arangod/Transaction/Methods.cpp [hunk 2]
```diff
@@ -1565,26 +1569,15 @@ Future<OperationResult> transaction::Methods::removeLocal(std::string const& col
...
-    } else if (value.isObject()) {
-      VPackSlice keySlice = value.get(StaticStrings::KeyString);
-      if (!keySlice.isString()) {
-        return Result(TRI_ERROR_ARANGO_DOCUMENT_HANDLE_BAD);
-      }
-      key = keySlice;
-    } else {
-      return Result(TRI_ERROR_ARANGO_DOCUMENT_HANDLE_BAD);
+
+    auto validatedKey = transaction::helpers::validatedOperationInputDocumentKey(*collection, value);
```
Comment (mchacki) [inline arangod/Transaction/Methods.cpp:1573]: This allows strings.
And objects with required _key string

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/10090/

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/10122/

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/10127/

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/10129/

File: arangod/Transaction/Helpers.cpp [hunk 3]
```diff
@@ -421,6 +424,51 @@ std::string transaction::helpers::makeIdFromCustom(CollectionNameResolver const*
...
+  }
+
+  if (collection.keyGenerator()->validateKey(keyRef.data(), keyRef.length())) {
+    return arangodb::ResultT<velocypack::StringRef>::success(keyRef);
+  }
+
+  return arangodb::ResultT<velocypack::StringRef>::error(TRI_ERROR_ARANGO_DOCUMENT_HANDLE_BAD);
+}
+
+VPackSlice transaction::helpers::normalizeRemoveOperationInput(LogicalCollection const& collection, VPackSlice const& value, VPackBuilder& builder) {
```
Comment (graetzer) [inline arangod/Transaction/Helpers.cpp:460]: this is potentially expensive

Comment (markuspf) [reply arangod/Transaction/Helpers.cpp:460]: explain.

**Pull request closed without merging**